# Assignment 5, Question 4: Data Exploration

**Points: 15**

In this notebook, you'll explore the clinical trial dataset using pandas selection and filtering techniques.

You'll use utility functions from `q3_data_utils` where helpful, but also demonstrate direct pandas operations.

## Setup

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Import utilities from Q3
from q3_data_utils import load_data, detect_missing, filter_data

# Load the data
df = load_data('data/clinical_trial_raw.csv')
print(f"Loaded {len(df)} patients with {len(df.columns)} variables")
display(df.head())


# Prewritten visualization functions for exploration (optional)
def plot_value_counts(series, title, figsize=(10, 6)):
    plt.figure(figsize=figsize)
    series.plot(kind='bar')
    plt.title(title)
    plt.xticks(rotation=45)
    plt.tight_layout()
    plt.show()

def plot_crosstab(crosstab_data, title, figsize=(10, 6)):
    plt.figure(figsize=figsize)
    plt.imshow(crosstab_data.values, cmap='Blues', aspect='auto')
    plt.colorbar()
    plt.title(title)
    plt.xticks(range(len(crosstab_data.columns)), crosstab_data.columns, rotation=45)
    plt.yticks(range(len(crosstab_data.index)), crosstab_data.index)
    plt.tight_layout()
    plt.show()


# REQUIRED: Site distribution + Save CSV
site_counts = df['site'].value_counts()
site_counts.to_csv("output/q4_site_counts.csv")
print("\n✅ Saved site counts to output/q4_site_counts.csv")
display(site_counts)


# REQUIRED: Summary statistics for numeric columns
print("\n Summary Statistics:")
display(df.describe())


# REQUIRED: Outlier detection (simple percentile check)
numeric_cols = df.select_dtypes(include=['float64', 'int64']).columns
print("\n Possible Outliers (1st–99th percentile):")
for col in numeric_cols:
    outliers = df[(df[col] < df[col].quantile(0.01)) | (df[col] > df[col].quantile(0.99))][col]
    print(f"{col}: {len(outliers)}")


# REQUIRED: Categorical value counts
print("\n Intervention Group Counts:")
display(df['intervention_group'].value_counts())

print("\n Sex Distribution:")
display(df['sex'].value_counts())


Loaded 10000 patients with 18 variables


patient_id  age         sex   bmi enrollment_date  systolic_bp  \
0     P00001   80           F  29.3      2022-05-01        123.0   
1     P00002   80    Female     NaN      2022-01-06        139.0   
2     P00003   82      Female  -1.0      2023-11-04        123.0   
3     P00004   95      Female  25.4      2022-08-15        116.0   
4     P00005   95           M   NaN      2023-04-17         97.0   

   diastolic_bp  cholesterol_total  cholesterol_hdl  cholesterol_ldl  \
0          80.0              120.0             55.0             41.0   
1          81.0              206.0             58.0            107.0   
2          86.0              172.0             56.0             82.0   
3          77.0              200.0             56.0            104.0   
4          71.0              185.0             78.0             75.0   

   glucose_fasting    site intervention_group  follow_up_months  \
0            118.0  site b            Control                20   
1             79.0  Site A            CONTROL                24   
2             77.0  SITE C      treatment b                   2   
3            115.0  Site D        treatment b                17   
4            113.0  site e         Treatmen A                 9   

   adverse_events outcome_cvd  adherence_pct dropout  
0               0          No           24.0      No  
1               0          No           77.0      No  
2               0         Yes           70.0      No  
3               0          No           62.0      No  
4               0         yes            NaN     Yes


✅ Saved site counts to output/q4_site_counts.csv


site
site b         742
Site B         736
SITE B         703
SITE A         684
Site  A        681
Site A         661
Site C         658
site a         651
site c         615
SITE C         605
Site D         362
site d         349
Site_D         332
Site E         319
SITE D         313
SITE E         295
site e         294
  SITE B        94
  site b        90
  Site B        88
  Site C        83
  site a        74
  SITE A        74
  Site  A       67
  Site A        64
  site c        57
  SITE C        55
  Site E        42
  SITE D        41
  site d        41
  site e        36
  Site D        32
  SITE E        31
  Site_D        31
Name: count, dtype: int64


 Summary Statistics:


age          bmi  systolic_bp  diastolic_bp  \
count  10000.000000  9562.000000  9586.000000   9586.000000   
mean      59.182700    25.730558   117.531087     73.550908   
std      151.769963     5.339547    13.973973     10.167464   
min     -999.000000    -1.000000    75.000000     60.000000   
25%       70.000000    23.500000   108.000000     65.000000   
50%       80.000000    26.000000   117.000000     73.000000   
75%       92.000000    28.775000   127.000000     81.000000   
max      100.000000    42.800000   173.000000    118.000000   

       cholesterol_total  cholesterol_hdl  cholesterol_ldl  glucose_fasting  \
count        9446.000000      9446.000000      9446.000000      9631.000000   
mean          178.039488        61.369786        85.698603        96.424255   
std            33.129034        11.062101        28.686463        17.112961   
min            91.000000        25.000000        40.000000        51.000000   
25%           155.000000        54.000000        65.000000        84.000000   
50%           177.000000        61.000000        84.000000        96.000000   
75%           200.000000        69.000000       105.000000       108.000000   
max           315.000000        98.000000       226.000000       163.000000   

       follow_up_months  adverse_events  adherence_pct  
count       10000.00000    10000.000000    8533.000000  
mean           12.25460        0.145500      60.615610  
std             7.07675        0.393631      18.974399  
min             0.00000        0.000000      20.000000  
25%             6.00000        0.000000      47.000000  
50%            12.00000        0.000000      62.000000  
75%            19.00000        0.000000      75.000000  
max            24.00000        4.000000     100.000000


 Possible Outliers (1st–99th percentile):
age: 0
bmi: 93
systolic_bp: 168
diastolic_bp: 95
cholesterol_total: 181
cholesterol_hdl: 158
cholesterol_ldl: 90
glucose_fasting: 176
follow_up_months: 0
adverse_events: 8
adherence_pct: 0

 Intervention Group Counts:


intervention_group
Contrl              802
TREATMENT B         761
Treatment  B        760
Control             751
treatment b         750
control             734
Treatment B         730
CONTROL             715
TreatmentA          635
Treatment A         610
Treatmen A          592
treatment a         588
TREATMENT A         572
  Treatment B       104
  Control            95
  treatment b        90
  CONTROL            84
  control            83
  TREATMENT B        83
  Contrl             73
  treatment a        67
  TreatmentA         66
  Treatment A        66
  TREATMENT A        65
  Treatment  B       64
  Treatmen A         60
Name: count, dtype: int64


 Sex Distribution:


sex
Female        2684
F             2340
M             2083
Male          1893
  Female       316
  F            247
  M            234
  Male         203
Name: count, dtype: int64

## Part 1: Basic Exploration (3 points)

Display:
1. Dataset shape
2. Column names and types
3. First 10 rows
4. Summary statistics (.describe())

In [2]:
# TODO: Display dataset info
# Display dataset info

import pandas as pd
import q3_data_utils as utils

# Load the dataset
df = utils.load_data("data/clinical_trial_raw.csv")

# 1. Dataset shape
print("Shape:", df.shape)

# 2. Column names and types
print("\nColumn Types:")
display(df.dtypes)

# 3. First 10 rows
print("\nFirst 10 rows:")
display(df.head(10))

# 4. Summary statistics
print("\nSummary Statistics:")
display(df.describe())


Shape: (10000, 18)

Column Types:


patient_id             object
age                     int64
sex                    object
bmi                   float64
enrollment_date        object
systolic_bp           float64
diastolic_bp          float64
cholesterol_total     float64
cholesterol_hdl       float64
cholesterol_ldl       float64
glucose_fasting       float64
site                   object
intervention_group     object
follow_up_months        int64
adverse_events          int64
outcome_cvd            object
adherence_pct         float64
dropout                object
dtype: object


First 10 rows:


patient_id  age         sex   bmi enrollment_date  systolic_bp  \
0     P00001   80           F  29.3      2022-05-01        123.0   
1     P00002   80    Female     NaN      2022-01-06        139.0   
2     P00003   82      Female  -1.0      2023-11-04        123.0   
3     P00004   95      Female  25.4      2022-08-15        116.0   
4     P00005   95           M   NaN      2023-04-17         97.0   
5     P00006   78           F  26.8      2023-08-29        116.0   
6     P00007   84           F  25.4      2022-05-12        133.0   
7     P00008   70        Male  24.7      2022-06-04        111.0   
8     P00009   92      Female  26.9      2022-04-06          NaN   
9     P00010   75        Male  21.1      2023-12-14        128.0   

   diastolic_bp  cholesterol_total  cholesterol_hdl  cholesterol_ldl  \
0          80.0              120.0             55.0             41.0   
1          81.0              206.0             58.0            107.0   
2          86.0              172.0             56.0             82.0   
3          77.0              200.0             56.0            104.0   
4          71.0              185.0             78.0             75.0   
5          66.0              164.0             54.0             99.0   
6         100.0              215.0             62.0            113.0   
7          72.0              174.0             60.0             94.0   
8           NaN              189.0             62.0             89.0   
9          76.0              218.0             77.0             97.0   

   glucose_fasting    site intervention_group  follow_up_months  \
0            118.0  site b            Control                20   
1             79.0  Site A            CONTROL                24   
2             77.0  SITE C      treatment b                   2   
3            115.0  Site D        treatment b                17   
4            113.0  site e         Treatmen A                 9   
5             99.0  Site A         TreatmentA                 4   
6             70.0  site a        treatment a                20   
7            109.0  SITE B        TREATMENT A                19   
8            103.0  site a            Control                21   
9             96.0  SITE A        Treatment B                 1   

   adverse_events outcome_cvd  adherence_pct dropout  
0               0          No           24.0      No  
1               0          No           77.0      No  
2               0         Yes           70.0      No  
3               0          No           62.0      No  
4               0         yes            NaN     Yes  
5               0         yes            NaN     Yes  
6               1          No           76.0      No  
7               0          No           53.0      No  
8               0         yes           53.0      No  
9               0          No           50.0      No


Summary Statistics:


age          bmi  systolic_bp  diastolic_bp  \
count  10000.000000  9562.000000  9586.000000   9586.000000   
mean      59.182700    25.730558   117.531087     73.550908   
std      151.769963     5.339547    13.973973     10.167464   
min     -999.000000    -1.000000    75.000000     60.000000   
25%       70.000000    23.500000   108.000000     65.000000   
50%       80.000000    26.000000   117.000000     73.000000   
75%       92.000000    28.775000   127.000000     81.000000   
max      100.000000    42.800000   173.000000    118.000000   

       cholesterol_total  cholesterol_hdl  cholesterol_ldl  glucose_fasting  \
count        9446.000000      9446.000000      9446.000000      9631.000000   
mean          178.039488        61.369786        85.698603        96.424255   
std            33.129034        11.062101        28.686463        17.112961   
min            91.000000        25.000000        40.000000        51.000000   
25%           155.000000        54.000000        65.000000        84.000000   
50%           177.000000        61.000000        84.000000        96.000000   
75%           200.000000        69.000000       105.000000       108.000000   
max           315.000000        98.000000       226.000000       163.000000   

       follow_up_months  adverse_events  adherence_pct  
count       10000.00000    10000.000000    8533.000000  
mean           12.25460        0.145500      60.615610  
std             7.07675        0.393631      18.974399  
min             0.00000        0.000000      20.000000  
25%             6.00000        0.000000      47.000000  
50%            12.00000        0.000000      62.000000  
75%            19.00000        0.000000      75.000000  
max            24.00000        4.000000     100.000000

## Part 2: Column Selection (3 points)

Demonstrate different selection methods:

1. Select only numeric columns using `.select_dtypes()`
2. Select specific columns by name
3. Select a subset of rows and columns using `.loc[]`

In [3]:
# TODO: Select numeric columns
# Select numeric columns
numeric_df = df.select_dtypes(include=['int64', 'float64'])
display(numeric_df.head())


age   bmi  systolic_bp  diastolic_bp  cholesterol_total  cholesterol_hdl  \
0   80  29.3        123.0          80.0              120.0             55.0   
1   80   NaN        139.0          81.0              206.0             58.0   
2   82  -1.0        123.0          86.0              172.0             56.0   
3   95  25.4        116.0          77.0              200.0             56.0   
4   95   NaN         97.0          71.0              185.0             78.0   

   cholesterol_ldl  glucose_fasting  follow_up_months  adverse_events  \
0             41.0            118.0                20               0   
1            107.0             79.0                24               0   
2             82.0             77.0                 2               0   
3            104.0            115.0                17               0   
4             75.0            113.0                 9               0   

   adherence_pct  
0           24.0  
1           77.0  
2           70.0  
3           62.0  
4            NaN

In [4]:
# TODO: Select specific columns
# Select specific columns by name
subset_cols = df[['patient_id', 'age', 'sex', 'site']]
display(subset_cols.head())


patient_id  age         sex    site
0     P00001   80           F  site b
1     P00002   80    Female    Site A
2     P00003   82      Female  SITE C
3     P00004   95      Female  Site D
4     P00005   95           M  site e

In [5]:
# TODO: Use .loc[] to select subset
# Select subset of rows and columns using .loc[]
subset_loc = df.loc[0:9, ['patient_id', 'age', 'intervention_group']]
display(subset_loc)


patient_id  age intervention_group
0     P00001   80            Control
1     P00002   80            CONTROL
2     P00003   82      treatment b  
3     P00004   95        treatment b
4     P00005   95         Treatmen A
5     P00006   78         TreatmentA
6     P00007   84        treatment a
7     P00008   70        TREATMENT A
8     P00009   92            Control
9     P00010   75        Treatment B

## Part 3: Filtering (4 points)

Filter the data to answer these questions:

1. How many patients are over 65 years old?
2. How many patients have systolic BP > 140?
3. Find patients who are both over 65 AND have systolic BP > 140
4. Find patients from Site A or Site B using `.isin()`

In [6]:
import q3_data_utils as utils
df = utils.load_data("data/clinical_trial_raw.csv")

# Filter and count patients over 65
age_filter = [{'column': 'age', 'condition': 'greater_than', 'value': 65}]
patients_over_65 = utils.filter_data(df, age_filter)
print(f"Patients over age 65: {len(patients_over_65)}")


Patients over age 65: 8326


In [7]:
import q3_data_utils as utils
df = utils.load_data("data/clinical_trial_raw.csv")
# Filter for systolic BP > 140
bp_filter = [{'column': 'systolic_bp', 'condition': 'greater_than', 'value': 140}]
high_bp = utils.filter_data(df, bp_filter)
print(f"Patients with systolic BP > 140: {len(high_bp)}")

Patients with systolic BP > 140: 538


In [8]:
import q3_data_utils as utils
df = utils.load_data("data/clinical_trial_raw.csv")
# Filter patients over 65 AND high systolic BP > 140
multi_filters = [
    {'column': 'age', 'condition': 'greater_than', 'value': 65},
    {'column': 'systolic_bp', 'condition': 'greater_than', 'value': 140}
]

both_conditions = utils.filter_data(df, multi_filters)
print(f"Patients over 65 AND systolic BP > 140: {len(both_conditions)}")



Patients over 65 AND systolic BP > 140: 464


In [9]:
# Clean 'site' column
df["site_clean"] = (
    df["site"]
    .str.lower()                # convert to lowercase
    .str.strip()                # remove leading/trailing spaces
    .str.replace(r"\s+", " ", regex=True)  # collapse multiple spaces into one
    .str.replace("_", " ")      # replace underscores with spaces
)



# Convert to title case ("site a" → "Site A")
df["site_clean"] = df["site_clean"].str.title()

# Verify
print(df["site_clean"].unique())
print("Number of unique sites:", df["site_clean"].nunique())


# Clean the 'intervention_group' column
df["intervention_group_clean"] = (
    df["intervention_group"]
    .str.strip()                  # remove leading/trailing spaces
    .str.lower()                  # convert to lowercase
    .replace(r"^con.*", "control", regex=True)   # anything starting with 'con' → 'control'
    .replace(r"^treatmen.*a.*", "treatment a", regex=True)  # variants like 'treatmen a', 'treatmenta' → 'treatment a'
    .replace(r"^treatmen.*b.*", "treatment b", regex=True)  # variants like 'treatmen b', 'treatmentb' → 'treatment b'
)

# Check cleaned unique values
unique_groups = df["intervention_group_clean"].unique()
print(unique_groups)

# Optional: count how many of each
print(df["intervention_group_clean"].value_counts())

['Site B' 'Site A' 'Site C' 'Site D' 'Site E']
Number of unique sites: 5


['control' 'treatment b' 'treatment a']
intervention_group_clean
treatment b    3342
control        3337
treatment a    3321
Name: count, dtype: int64


In [10]:
# Drop old columns and replace them with cleaned versions
df = df.drop(columns=["site", "intervention_group"], errors="ignore")

# Rename cleaned columns to original names
df = df.rename(columns={
    "site_clean": "site",
    "intervention_group_clean": "intervention_group"
})

In [11]:
import q3_data_utils as utils
df = utils.load_data("data/clinical_trial_raw.csv")
# Filter patients from Site A or Site B
site_filter = [{'column': 'site', 'condition': 'in_list', 'value': ['Site A', 'Site B']}]

site_ab = utils.filter_data(df, site_filter)
print(f"Patients from Site A or Site B: {len(site_ab)}")


Patients from Site A or Site B: 1397


## Part 4: Value Counts and Grouping (5 points)

1. Get value counts for the 'site' column
2. Get value counts for the 'intervention_group' column  
3. Create a crosstab of site vs intervention_group
4. Calculate mean age by site
5. Save the site value counts to `output/q4_site_counts.csv`

In [12]:
import pandas as pd  # ✅ ensure pandas is imported

# 1. Value counts for site
site_counts = df['site'].value_counts()
print("Site value counts:")
display(site_counts)

# 2. Value counts for intervention groups
print("\nIntervention group counts:")
display(df['intervention_group'].value_counts())

# 3. Crosstab: site vs intervention_group
print("\nCrosstab: site vs intervention group")
site_intervention_crosstab = pd.crosstab(df['site'], df['intervention_group'])
display(site_intervention_crosstab)

# 4. Mean age by site
print("\nMean age by site:")
mean_age_by_site = df.groupby('site')['age'].mean()
display(mean_age_by_site)


Site value counts:


site
site b         742
Site B         736
SITE B         703
SITE A         684
Site  A        681
Site A         661
Site C         658
site a         651
site c         615
SITE C         605
Site D         362
site d         349
Site_D         332
Site E         319
SITE D         313
SITE E         295
site e         294
  SITE B        94
  site b        90
  Site B        88
  Site C        83
  site a        74
  SITE A        74
  Site  A       67
  Site A        64
  site c        57
  SITE C        55
  Site E        42
  SITE D        41
  site d        41
  site e        36
  Site D        32
  SITE E        31
  Site_D        31
Name: count, dtype: int64


Intervention group counts:


intervention_group
Contrl              802
TREATMENT B         761
Treatment  B        760
Control             751
treatment b         750
control             734
Treatment B         730
CONTROL             715
TreatmentA          635
Treatment A         610
Treatmen A          592
treatment a         588
TREATMENT A         572
  Treatment B       104
  Control            95
  treatment b        90
  CONTROL            84
  control            83
  TREATMENT B        83
  Contrl             73
  treatment a        67
  TreatmentA         66
  Treatment A        66
  TREATMENT A        65
  Treatment  B       64
  Treatmen A         60
Name: count, dtype: int64


Crosstab: site vs intervention group


intervention_group    CONTROL      Contrl      Control      TREATMENT A    \
site                                                                        
  SITE A                      0           0            0                0   
  SITE B                      0           0            1                0   
  SITE C                      0           0            0                1   
  SITE D                      0           0            0                0   
  SITE E                      0           0            0                0   
  Site  A                     1           0            1                0   
  Site A                      1           0            0                0   
  Site B                      3           2            0                0   
  Site C                      0           0            1                0   
  Site D                      0           0            0                0   
  Site E                      0           0            1                0   
  Site_D                      0           0            0                1   
  site a                      0           2            0                1   
  site b                      1           1            0                0   
  site c                      0           0            0                2   
  site d                      0           1            0                0   
  site e                      0           0            1                1   
SITE A                        3           2            6                2   
SITE B                        6           4           10                4   
SITE C                        4           2            9                8   
SITE D                        6           2            0                2   
SITE E                        1           1            4                4   
Site  A                       3           9            3                1   
Site A                        9           7            3                6   
Site B                        5          11           11                7   
Site C                        6           6            4                5   
Site D                        3           2            5                1   
Site E                        5           4            6                3   
Site_D                        2           3            1                2   
site a                        7           5            5                3   
site b                        6           2            8                5   
site c                        6           4            8                4   
site d                        4           2            6                1   
site e                        2           1            1                1   

intervention_group    TREATMENT B      Treatmen A      Treatment  B    \
site                                                                    
  SITE A                          1               0                 0   
  SITE B                          1               1                 2   
  SITE C                          0               0                 0   
  SITE D                          1               0                 0   
  SITE E                          0               0                 0   
  Site  A                         0               0                 1   
  Site A                          0               1                 0   
  Site B                          1               0                 0   
  Site C                          1               0                 0   
  Site D                          0               0                 0   
  Site E                          1               0                 0   
  Site_D                          0               0                 0   
  site a                          0               0                 1   
  site b                          0               1                 0   
  site c                          0               0        


Mean age by site:


site
  SITE A       23.608108
  SITE B       59.670213
  SITE C       60.254545
  SITE D       77.853659
  SITE E       12.258065
  Site  A      80.328358
  Site A       49.843750
  Site B       56.318182
  Site C       27.831325
  Site D      -25.312500
  Site E       75.500000
  Site_D        9.870968
  site a       67.270270
  site b       57.022222
  site c       80.929825
  site d       81.000000
  site e       83.194444
SITE A         60.910819
SITE B         63.628734
SITE C         55.459504
SITE D         55.511182
SITE E         51.867797
Site  A        53.770925
Site A         61.440242
Site B         56.911685
Site C         60.890578
Site D         67.472376
Site E         56.429467
Site_D         77.993976
site a         49.803379
site b         68.405660
site c         56.832520
site d         61.882521
site e         61.860544
Name: age, dtype: float64

In [13]:
# Save site value counts to output folder
site_counts.to_csv("output/q4_site_counts.csv")
print("✅ Saved site counts to: output/q4_site_counts.csv")


✅ Saved site counts to: output/q4_site_counts.csv


## Summary

Write 2-3 sentences about what you learned from exploring this dataset.

From the dataset exploration, I learned that we have 10,000 patients and 18 variables with a mix of numeric and categorical data types. There are missing values coded as NaN as well as incorrect sentinel entries (e.g., BMI = -1) and inconsistent text formatting in columns such as sex, site, and intervention_group. Summary statistics showed mean age ≈ 59 years and mean BMI ≈ 25.7, and filtering demonstrated meaningful subsets like 8326 patients >65 years and 538 with elevated systolic BP >140.

**Your summary here:**

TODO: Write your observations


The dataset contains 10,000 patients and 18 variables with mixed data types, including demographics, clinical measures, and intervention details. Exploration revealed missing values and incorrect entries such as sentinel values (e.g., BMI = -1) as well as inconsistent text formatting in categorical fields like sex, site, and intervention_group. Summary statistics show an average age of around 59 years and BMI of ~25.7, with subsets such as 8326 patients over 65 years and 538 patients with elevated systolic blood pressure (>140 mmHg).